### PR TITLE
[core][autoscaler][v2] Autoscaler instance 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR adds the AutoscalerInstance data structure that represents an autoscaler instance with 2 states:

the ray state from GCS
the Instance Manager (IM) instance state.
This is a key data structure for reconcilation between the two worlds.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                # This happens when the a launch request is made to the node provider.
+                Instance.REQUESTED
+            },
+            # When in this status, a launch request to the node provider is made.
+            Instance.REQUESTED: {
+                # Cloud provider allocated a cloud instance for the instance.
+                # This happens when the cloud instance first appears in the list of
+                # running cloud instances from the cloud instance provider.
+                Instance.ALLOCATED,
+                # Retry the allocation, become queueing again.
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/common.py:146]: This is to transition directly to QUEUED for retry.

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+                Instance.STOPPED
+            },
+            # Whenever a cloud instance disappears from the list of running cloud
+            # instances from the node provider, the instance is marked as stopped. Since
+            # we guarantee 1:1 mapping of a Instance to a cloud instance, this is a
+            # terminal state.
+            Instance.STOPPED: set(),  # Terminal state.
+            # When in this status, the cloud instance failed to be allocated by the
+            # node provider.
+            Instance.ALLOCATION_FAILED: set(),  # Terminal state.
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/common.py:235]: ALLOCATION_FAILED is now a terminal instance state.

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            Instance.REQUESTED: {
+                # Cloud provider allocated a cloud instance for the instance.
+                # This happens when the cloud instance first appears in the list of
+                # running cloud instances from the cloud instance provider.
+                Instance.ALLOCATED,
+                # Retry the allocation, become queueing again.
+                Instance.QUEUED,
+                # Cloud provider fails to allocate one. Either as a timeout or
+                # the launch request fails immediately.
+                Instance.ALLOCATION_FAILED,
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/common.py:150]: Do we have a retry count?

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                Instance.STOPPING,
+                # cloud instance somehow failed.
+                Instance.STOPPED,
+            },
+            # Initial state before the instance is created. Should never be used.
+            Instance.UNKNOWN: set(),
+        }
+
+    @staticmethod
+    def get_status_times_ns(
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/common.py:247]: nit: get_status_transition_times_ns

**Review by jjyao — APPROVED** [review]

File: python/ray/autoscaler/v2/instance_manager/common.py (hunk 3 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/common.py:150]: i was thinking of using the number of `QUEUED` timestamps on the status transitions for this. But I am open to having a field tracking it explicitly.

**Pull request merged**